[BUG] fastparquet_compatibility_test fails on dataproc

fastparquet_compatibility_test failed in a recent nightly Dataproc build:
```
[2023-11-01T16:49:29.309Z] FAILED ../../src/main/python/fastparquet_compatibility_test.py::test_reading_file_rewritten_with_fastparquet[Date(not_null)-int961] - IsADirectoryError: [Errno 21] Is a directory: '/tmp/pyspark_tests//rapids-i...
[2023-11-01T16:49:29.309Z] FAILED ../../src/main/python/fastparquet_compatibility_test.py::test_reading_file_rewritten_with_fastparquet[Date-int96][INJECT_OOM] - IsADirectoryError: [Errno 21] Is a directory: '/tmp/pyspark_tests//rapids-i...
[2023-11-01T16:49:29.309Z] FAILED ../../src/main/python/fastparquet_compatibility_test.py::test_reading_file_rewritten_with_fastparquet[Timestamp(not_null)-int960] - IsADirectoryError: [Errno 21] Is a directory: '/tmp/pyspark_tests//rapids-i...
[2023-11-01T16:49:29.309Z] FAILED ../../src/main/python/fastparquet_compatibility_test.py::test_reading_file_rewritten_with_fastparquet[Timestamp-int96] - IsADirectoryError: [Errno 21] Is a directory: '/tmp/pyspark_tests//rapids-i...\
[2023-11-01T16:49:29.309Z] FAILED ../../src/main/python/fastparquet_compatibility_test.py::test_reading_file_rewritten_with_fastparquet[Struct(not_null)(('first', Integer(not_null)))-int96][INJECT_OOM] - IsADirectoryError: [Errno 21] Is a directory: '/tmp/pyspark_tests//rapids-i...
```

<details closed>
<summary>Details for one of the test failures</summary>

```
[2023-11-01T16:49:29.050Z] =================================== FAILURES ===================================
[2023-11-01T16:49:29.050Z] _____ test_reading_file_rewritten_with_fastparquet[Date(not_null)-int961] ______
[2023-11-01T16:49:29.050Z] 
[2023-11-01T16:49:29.050Z] self = <[AttributeError("'ParquetFile' object has no attribute '_schema'") raised in repr()] ParquetFile object at 0x7fdf20583d00>
[2023-11-01T16:49:29.050Z] fn = '/tmp/pyspark_tests//rapids-it-dataproc-20-ubuntu18-322-m-main-11369-1843522649/_local/FASTPARQUET_WRITE_PATH_tmp'
[2023-11-01T16:49:29.050Z] verify = False, open_with = <function default_open at 0x7fdf309460d0>
[2023-11-01T16:49:29.050Z] root = False, sep = None
[2023-11-01T16:49:29.050Z] 
[2023-11-01T16:49:29.050Z]     def __init__(self, fn, verify=False, open_with=default_open,
[2023-11-01T16:49:29.050Z]                  root=False, sep=None):
[2023-11-01T16:49:29.050Z]         if isinstance(fn, (tuple, list)):
[2023-11-01T16:49:29.050Z]             basepath, fmd = metadata_from_many(fn, verify_schema=verify,
[2023-11-01T16:49:29.050Z]                                                open_with=open_with, root=root)
[2023-11-01T16:49:29.050Z]             if basepath:
[2023-11-01T16:49:29.050Z]                 self.fn = join_path(basepath, '_metadata')  # effective file
[2023-11-01T16:49:29.050Z]             else:
[2023-11-01T16:49:29.050Z]                 self.fn = '_metadata'
[2023-11-01T16:49:29.050Z]             self.fmd = fmd
[2023-11-01T16:49:29.050Z]             self._set_attrs()
[2023-11-01T16:49:29.050Z]         elif hasattr(fn, 'read'):
[2023-11-01T16:49:29.050Z]             # file-like
[2023-11-01T16:49:29.050Z]             self._parse_header(fn, verify)
[2023-11-01T16:49:29.050Z]             if self.file_scheme not in ['simple', 'empty']:
[2023-11-01T16:49:29.050Z]                 raise ValueError('Cannot use file-like input '
[2023-11-01T16:49:29.050Z]                                  'with multi-file data')
[2023-11-01T16:49:29.050Z]             open_with = lambda *args, **kwargs: fn
[2023-11-01T16:49:29.050Z]             self.fn = None
[2023-11-01T16:49:29.050Z]         else:
[2023-11-01T16:49:29.050Z]             try:
[2023-11-01T16:49:29.050Z]                 fn2 = join_path(fn, '_metadata')
[2023-11-01T16:49:29.050Z]                 self.fn = fn2
[2023-11-01T16:49:29.050Z] >               with open_with(fn2, 'rb') as f:
[2023-11-01T16:49:29.050Z] 
[2023-11-01T16:49:29.050Z] /opt/conda/default/lib/python3.8/site-packages/fastparquet/api.py:110: 
[2023-11-01T16:49:29.050Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2023-11-01T16:49:29.050Z] 
[2023-11-01T16:49:29.050Z] f = '/tmp/pyspark_tests//rapids-it-dataproc-20-ubuntu18-322-m-main-11369-1843522649/_local/FASTPARQUET_WRITE_PATH_tmp/_metadata'
[2023-11-01T16:49:29.050Z] mode = 'rb'
[2023-11-01T16:49:29.050Z] 
[2023-11-01T16:49:29.050Z]     def default_open(f, mode='rb'):
[2023-11-01T16:49:29.050Z] >       return open(f, mode)
[2023-11-01T16:49:29.050Z] E       FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pyspark_tests//rapids-it-dataproc-20-ubuntu18-322-m-main-11369-1843522649/_local/FASTPARQUET_WRITE_PATH_tmp/_metadata'
[2023-11-01T16:49:29.050Z] 
[2023-11-01T16:49:29.050Z] /opt/conda/default/lib/python3.8/site-packages/fastparquet/util.py:39: FileNotFoundError
[2023-11-01T16:49:29.050Z] 
[2023-11-01T16:49:29.050Z] During handling of the above exception, another exception occurred:
[2023-11-01T16:49:29.050Z] 
[2023-11-01T16:49:29.050Z] column_gen = Date(not_null), time_format = 'int96'
[2023-11-01T16:49:29.050Z] spark_tmp_path = '/tmp/pyspark_tests//rapids-it-dataproc-20-ubuntu18-322-m-main-11369-1843522649/'
[2023-11-01T16:49:29.050Z] 
[2023-11-01T16:49:29.050Z]     @pytest.mark.skipif(condition=fastparquet_unavailable(),
[2023-11-01T16:49:29.050Z]                         reason="fastparquet is required for testing fastparquet compatibility")
[2023-11-01T16:49:29.050Z]     @pytest.mark.parametrize('column_gen, time_format', [
[2023-11-01T16:49:29.050Z]         pytest.param(
[2023-11-01T16:49:29.050Z]             DateGen(nullable=False,
[2023-11-01T16:49:29.050Z]                      start=pandas_min_date,
[2023-11-01T16:49:29.050Z]                      end=pandas_max_date), 'int64',
[2023-11-01T16:49:29.050Z]             marks=pytest.mark.xfail(reason="Apache Spark and the plugin both have problems reading dates written via "
[2023-11-01T16:49:29.050Z]                                            "fastparquet, if written in int64: "
[2023-11-01T16:49:29.050Z]                                            "\"Illegal Parquet type: INT64 (TIMESTAMP(NANOS,false)).\"")),
[2023-11-01T16:49:29.050Z]         pytest.param(
[2023-11-01T16:49:29.050Z]             DateGen(nullable=False), 'int96',
[2023-11-01T16:49:29.050Z]             marks=pytest.mark.xfail(reason="fastparquet does not support int96RebaseModeInWrite, for dates before "
[2023-11-01T16:49:29.050Z]                                            "1582-10-15 or timestamps before 1900-01-01T00:00:00Z. "
[2023-11-01T16:49:29.050Z]                                            "This messes up reads from Apache Spark and the plugin.")),
[2023-11-01T16:49:29.050Z]         (DateGen(nullable=False,
[2023-11-01T16:49:29.050Z]                  start=date(year=2000, month=1, day=1),
[2023-11-01T16:49:29.050Z]                  end=date(year=2020, month=12, day=31)), 'int96'),
[2023-11-01T16:49:29.050Z]         (DateGen(nullable=True,
[2023-11-01T16:49:29.050Z]                 start=date(year=2000, month=1, day=1),
[2023-11-01T16:49:29.050Z]                 end=date(year=2020, month=12, day=31)), 'int96'),
[2023-11-01T16:49:29.050Z]         pytest.param(
[2023-11-01T16:49:29.050Z]             TimestampGen(nullable=False,
[2023-11-01T16:49:29.050Z]                 start=datetime(2000, 1, 1, tzinfo=timezone.utc),
[2023-11-01T16:49:29.050Z]                 end=datetime(2200, 12, 31, tzinfo=timezone.utc)), 'int64',
[2023-11-01T16:49:29.050Z]             marks=pytest.mark.xfail(reason="Apache Spark and the plugin both have problems reading timestamps written via "
[2023-11-01T16:49:29.050Z]                                            "fastparquet, if written in int64: "
[2023-11-01T16:49:29.050Z]                                            "\"Illegal Parquet type: INT64 (TIMESTAMP(NANOS,false)).\"")),
[2023-11-01T16:49:29.050Z]         (TimestampGen(nullable=False,
[2023-11-01T16:49:29.050Z]                      start=datetime(2000, 1, 1, tzinfo=timezone.utc),
[2023-11-01T16:49:29.050Z]                      end=datetime(2200, 12, 31, tzinfo=timezone.utc)), 'int96'),
[2023-11-01T16:49:29.050Z]         (TimestampGen(nullable=True,
[2023-11-01T16:49:29.050Z]                       start=datetime(2000, 1, 1, tzinfo=timezone.utc),
[2023-11-01T16:49:29.050Z]                       end=datetime(2200, 12, 31, tzinfo=timezone.utc)), 'int96'),
[2023-11-01T16:49:29.050Z]         pytest.param(
[2023-11-01T16:49:29.050Z]             TimestampGen(nullable=False), 'int96',
[2023-11-01T16:49:29.050Z]             marks=pytest.mark.xfail(reason="fastparquet does not support int96RebaseModeInWrite, for dates before "
[2023-11-01T16:49:29.051Z]                                            "1582-10-15 or timestamps before 1900-01-01T00:00:00Z. "
[2023-11-01T16:49:29.051Z]                                            "This messes up reads from Apache Spark and the plugin.")),
[2023-11-01T16:49:29.051Z]         pytest.param(
[2023-11-01T16:49:29.051Z]             ArrayGen(nullable=False, child_gen=IntegerGen(nullable=False)), 'int96',
[2023-11-01T16:49:29.051Z]             marks=pytest.mark.xfail(reason="fastparquet fails to serialize array elements with any available encoding. "
[2023-11-01T16:49:29.051Z]                                            "E.g. \"Error converting column 'a' to bytes using encoding JSON. "
[2023-11-01T16:49:29.051Z]                                            "Original error: Object of type int32 is not JSON serializable\".")),
[2023-11-01T16:49:29.051Z]         (StructGen(nullable=False, children=[('first', IntegerGen(nullable=False))]), 'int96'),
[2023-11-01T16:49:29.051Z]         pytest.param(
[2023-11-01T16:49:29.051Z]             StructGen(nullable=True, children=[('first', IntegerGen(nullable=False))]), 'int96',
[2023-11-01T16:49:29.051Z]             marks=pytest.mark.xfail(reason="fastparquet fails to read nullable Struct columns written from Apache Spark. "
[2023-11-01T16:49:29.051Z]                                            "It fails the rewrite to parquet, thereby failing the test.")),
[2023-11-01T16:49:29.051Z]     ], ids=idfn)
[2023-11-01T16:49:29.051Z]     def test_reading_file_rewritten_with_fastparquet(column_gen, time_format, spark_tmp_path):
[2023-11-01T16:49:29.051Z]         """
[2023-11-01T16:49:29.051Z]         This test is a workaround to test data-types that have problems being converted
[2023-11-01T16:49:29.051Z]         from Spark dataframes to Pandas dataframes.
[2023-11-01T16:49:29.051Z]         For instance, sparkDF.toPandas() incorrectly converts ARRAY<INT> columns into
[2023-11-01T16:49:29.051Z]         STRING columns.
[2023-11-01T16:49:29.051Z]         This test writes the Spark dataframe into a temporary file, and then uses
[2023-11-01T16:49:29.051Z]         `fastparquet` to read and write the file again, to the final destination.
[2023-11-01T16:49:29.051Z]         The final file should be in the correct format, with the right datatypes.
[2023-11-01T16:49:29.051Z]         This is then checked for read-accuracy, via CPU and GPU.
[2023-11-01T16:49:29.051Z]         """
[2023-11-01T16:49:29.051Z]         suffix = "/FASTPARQUET_WRITE_PATH"
[2023-11-01T16:49:29.051Z]         hdfs_data_path = spark_tmp_path + suffix
[2023-11-01T16:49:29.051Z]         local_base_path = (spark_tmp_path + "_local")
[2023-11-01T16:49:29.051Z]         local_data_path = local_base_path + suffix
[2023-11-01T16:49:29.051Z]     
[2023-11-01T16:49:29.051Z]         def rewrite_with_fastparquet(spark, data_gen):
[2023-11-01T16:49:29.051Z]             """
[2023-11-01T16:49:29.051Z]             This helper function (eventually) writes data generated from `data_gen` to the local filesystem,
[2023-11-01T16:49:29.051Z]             via fastparquet.
[2023-11-01T16:49:29.051Z]             To preserve data types from data_gen, the writes are done first through Spark, thus:
[2023-11-01T16:49:29.051Z]               1. Write to HDFS with Spark. (This preserves data types.)
[2023-11-01T16:49:29.051Z]               2. Copy the data to local FS (so that fastparquet can read it).
[2023-11-01T16:49:29.051Z]               3. Read data with fastparquet, write it back out to local FS with fastparquet.
[2023-11-01T16:49:29.051Z]               4. Copy the fastparquet output back to HDFS, for reads with Spark.
[2023-11-01T16:49:29.051Z]             """
[2023-11-01T16:49:29.051Z]             import fastparquet
[2023-11-01T16:49:29.051Z]             hdfs_tmp_data_path = hdfs_data_path + "_tmp"
[2023-11-01T16:49:29.051Z]             spark_df = gen_df(spark, data_gen, 2048)
[2023-11-01T16:49:29.051Z]             # 1. Write to HDFS with Spark.
[2023-11-01T16:49:29.051Z]             spark_df.repartition(1).write.mode("overwrite").parquet(hdfs_tmp_data_path)
[2023-11-01T16:49:29.051Z]             # Make local base directory.
[2023-11-01T16:49:29.051Z]             os.makedirs(name=local_base_path, exist_ok=True)
[2023-11-01T16:49:29.051Z]             # 2. Copy Spark-written data to local filesystem, for read with Parquet.
[2023-11-01T16:49:29.051Z]             local_tmp_data_path = local_data_path + "_tmp"
[2023-11-01T16:49:29.051Z]             copy_to_local(spark=spark, hdfs_source=hdfs_tmp_data_path, local_target=local_tmp_data_path)
[2023-11-01T16:49:29.051Z]             # 3. Read local tmp data with fastparquet, rewrite to final local path, with fastparquet.
[2023-11-01T16:49:29.051Z]             pandas_df = fastparquet.ParquetFile(local_tmp_data_path).to_pandas()
[2023-11-01T16:49:29.051Z]             fastparquet.write(local_data_path, pandas_df, times=time_format)
[2023-11-01T16:49:29.051Z]             # 4. Copy fastparquet-written data back to HDFS, so that Spark can read it.
[2023-11-01T16:49:29.051Z]             copy_from_local(spark=spark, local_source=local_data_path, hdfs_target=hdfs_data_path)
[2023-11-01T16:49:29.051Z]             # Local data can now be cleaned up.
[2023-11-01T16:49:29.051Z]             delete_local_directory(local_base_path)
[2023-11-01T16:49:29.051Z]     
[2023-11-01T16:49:29.051Z]         gen = StructGen([('a', column_gen),
[2023-11-01T16:49:29.051Z]                          ('part', IntegerGen(nullable=False))], nullable=False)
[2023-11-01T16:49:29.051Z]         # Write data with CPU session.
[2023-11-01T16:49:29.051Z] >       with_cpu_session(
[2023-11-01T16:49:29.051Z]             lambda spark: rewrite_with_fastparquet(spark, gen)
[2023-11-01T16:49:29.051Z]         )
[2023-11-01T16:49:29.051Z] 
[2023-11-01T16:49:29.051Z] ../../src/main/python/fastparquet_compatibility_test.py:428: 
[2023-11-01T16:49:29.051Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2023-11-01T16:49:29.051Z] ../../src/main/python/spark_session.py:116: in with_cpu_session
[2023-11-01T16:49:29.051Z]     return with_spark_session(func, conf=copy)
[2023-11-01T16:49:29.051Z] ../../src/main/python/spark_session.py:100: in with_spark_session
[2023-11-01T16:49:29.051Z]     ret = func(_spark)
[2023-11-01T16:49:29.051Z] ../../src/main/python/fastparquet_compatibility_test.py:429: in <lambda>
[2023-11-01T16:49:29.051Z]     lambda spark: rewrite_with_fastparquet(spark, gen)
[2023-11-01T16:49:29.051Z] ../../src/main/python/fastparquet_compatibility_test.py:418: in rewrite_with_fastparquet
[2023-11-01T16:49:29.051Z]     pandas_df = fastparquet.ParquetFile(local_tmp_data_path).to_pandas()
[2023-11-01T16:49:29.051Z] /opt/conda/default/lib/python3.8/site-packages/fastparquet/api.py:115: in __init__
[2023-11-01T16:49:29.051Z]     with open_with(fn, 'rb') as f:
[2023-11-01T16:49:29.051Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2023-11-01T16:49:29.051Z] 
[2023-11-01T16:49:29.051Z] f = '/tmp/pyspark_tests//rapids-it-dataproc-20-ubuntu18-322-m-main-11369-1843522649/_local/FASTPARQUET_WRITE_PATH_tmp'
[2023-11-01T16:49:29.051Z] mode = 'rb'
[2023-11-01T16:49:29.051Z] 
[2023-11-01T16:49:29.051Z]     def default_open(f, mode='rb'):
[2023-11-01T16:49:29.051Z] >       return open(f, mode)
[2023-11-01T16:49:29.051Z] E       IsADirectoryError: [Errno 21] Is a directory: '/tmp/pyspark_tests//rapids-it-dataproc-20-ubuntu18-322-m-main-11369-1843522649/_local/FASTPARQUET_WRITE_PATH_tmp'
[2023-11-01T16:49:29.051Z] 
[2023-11-01T16:49:29.051Z] /opt/conda/default/lib/python3.8/site-packages/fastparquet/util.py:39: IsADirectoryError
```
</details>

--- a/integration_tests/requirements.txt
+++ b/integration_tests/requirements.txt
@@ -17,4 +17,4 @@ pandas
 pyarrow
 pytest-xdist >= 2.0.0
 findspark
-fastparquet >= 0.8.3
+fastparquet == 0.8.3

--- a/integration_tests/requirements.txt
+++ b/integration_tests/requirements.txt
@@ -17,4 +17,4 @@ pandas
 pyarrow
 pytest-xdist >= 2.0.0
 findspark
-fastparquet >= 2023.8.0
+fastparquet >= 0.8.3

--- a/integration_tests/src/assembly/bin.xml
+++ b/integration_tests/src/assembly/bin.xml
@@ -43,6 +43,10 @@
             <outputDirectory>integration_tests</outputDirectory>
         </file>
         <file>
+            <source>${spark.rapids.source.basedir}/integration_tests/requirements.txt</source>
+            <outputDirectory>integration_tests</outputDirectory>
+        </file>
+        <file>
             <source>${project.build.directory}/extra-resources/rapids4spark-version-info.properties</source>
             <outputDirectory>integration_tests</outputDirectory>
         </file>

--- a/jenkins/Dockerfile-blossom.integration.rocky
+++ b/jenkins/Dockerfile-blossom.integration.rocky
@@ -57,7 +57,7 @@ RUN export CUDA_VER=`echo ${CUDA_VER} | cut -d '.' -f 1,2` && \
     conda install -y -c conda-forge sre_yield && \
     conda clean -ay
 # install pytest plugins for xdist parallel run
-RUN python -m pip install findspark pytest-xdist pytest-order fastparquet
+RUN python -m pip install findspark pytest-xdist pytest-order fastparquet==0.8.3
 
 # Set default java as 1.8.0
 ENV JAVA_HOME "/usr/lib/jvm/java-1.8.0-openjdk"

--- a/jenkins/Dockerfile-blossom.integration.ubuntu
+++ b/jenkins/Dockerfile-blossom.integration.ubuntu
@@ -69,7 +69,7 @@ RUN export CUDA_VER=`echo ${CUDA_VER} | cut -d '.' -f 1,2` && \
     conda install -y -c conda-forge sre_yield && \
     conda clean -ay
 # install pytest plugins for xdist parallel run
-RUN python -m pip install findspark pytest-xdist pytest-order fastparquet
+RUN python -m pip install findspark pytest-xdist pytest-order fastparquet==0.8.3
 
 RUN apt install -y inetutils-ping expect
 

--- a/jenkins/Dockerfile-blossom.ubuntu
+++ b/jenkins/Dockerfile-blossom.ubuntu
@@ -58,7 +58,7 @@ RUN update-java-alternatives --set /usr/lib/jvm/java-1.8.0-openjdk-amd64
 
 RUN ln -sfn /usr/bin/python3.8 /usr/bin/python
 RUN ln -sfn /usr/bin/python3.8 /usr/bin/python3
-RUN python -m pip install pytest sre_yield requests pandas pyarrow findspark pytest-xdist pre-commit pytest-order fastparquet
+RUN python -m pip install pytest sre_yield requests pandas pyarrow findspark pytest-xdist pre-commit pytest-order fastparquet==0.8.3
 
 # libnuma1 and libgomp1 are required by ucx packaging
 RUN apt install -y inetutils-ping expect wget libnuma1 libgomp1

--- a/jenkins/databricks/setup.sh
+++ b/jenkins/databricks/setup.sh
@@ -53,4 +53,4 @@ $PYSPARK_PYTHON -m pip install --target $PYTHON_SITE_PACKAGES pytest sre_yield r
 
 # Install fastparquet (and numpy as its dependency).
 $PYSPARK_PYTHON -m pip install --target $PYTHON_SITE_PACKAGES numpy
-$PYSPARK_PYTHON -m pip install --target $PYTHON_SITE_PACKAGES fastparquet
+$PYSPARK_PYTHON -m pip install --target $PYTHON_SITE_PACKAGES fastparquet==0.8.3

